crates/runtime/src/task_history/otel_exporter.rs

```diff
@@ -147,10 +170,27 @@ impl TaskHistoryExporter {
 
 impl SpanExporter for TaskHistoryExporter {
     fn export(&mut self, batch: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
-        let spans = batch
+        let mut spans: Vec<TaskSpan> = batch
             .into_iter()
             .map(|span| self.span_to_task_span(span))
             .collect();
+
+        // If any span has defined a [`TaskSpan::trace_id_override`], we must override others.
+        let overrides: HashMap<Arc<str>, Arc<str>> = spans
+            .iter()
+            .filter_map(|span| {
+                span.trace_id_override
+                    .as_ref()
+                    .map(|new_trace| (Arc::clone(&span.trace_id), Arc::clone(new_trace)))
+            })
+            .collect();
+
+        for span in &mut spans {
+            if let Some(new_trace_id) = overrides.get(&span.trace_id) {
+                span.trace_id = Arc::clone(new_trace_id);
+            }
+        }
```

Comment:
Why are we doing a double iteration over the spans here? Couldn't we just directly apply the override with a single loop?

Reply:
From my understanding of how we accumulate the spans, if 2 or more spans are in a trace, only 1 of them may have the override. We then need to apply the override to the remaining >1 spans.